workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/InterpolateTransformSystem.cs

```diff
@@ -1,127 +1,190 @@
+            public ComponentDataArray<TicksSinceLastUpdate> TicksSinceLastUpdate;
+            [ReadOnly] public ComponentDataArray<TransformInternal.ReceivedUpdates> Updates;
+            [ReadOnly] public ComponentDataArray<TransformInternal.Component> CurrentTransform;
+            [ReadOnly] public ComponentDataArray<NotAuthoritative<TransformInternal.Component>> DenotesNotAuthoritative;
         }
 
-        [Inject] private TransformData transformData;
-
-        protected override void OnCreateManager(int capacity)
-        {
-            base.OnCreateManager(capacity);
-
-            origin = World.GetExistingManager<WorkerSystem>().Origin;
+        [Inject] private Data data;
+        [Inject] private TickRateEstimationSystem tickRateSystem;
 
-            tickSystem = World.GetOrCreateManager<TickSystem>();
-        }
-
-        /*
-         * This system receives transform updates from the server and applies them on the client.
-         * Updates are not applied immedately due to the network sending updates at an inconsistent rate
-         * By keeping a buffer of updates, motion can still look smooth at the cost of being
-         * slightly behind the true position of the object on the server.
-         *
-         * The strategy used is:
-         * 1. Drop any updates that are too far in the past
-         * 2. If the time the update is supposed to be applied matches the client tick time, apply it.
-         * 3. If the next update is supposed to be applied in the future, interpolate the position for the
-         * current tick using the last position and the next update.
-         *
-         * The Unity ECS Transform component contains the latest transfrom update and does not accurately
-         * reflect the position of the object in the present. The rigid body transform is the true rendered
-         * transform of the object.
-         */
+        // todo this does everything eagerly, but with smearing it will probably make more sense to do things lazily
+        // the main difference would be how to detect the buffer is full and needs to be emptied
         protected override void OnUpdate()
         {
-            for (var i = 0; i < transformData.Length; i++)
+            for (int i = 0; i < data.Length; ++i)
             {
-                var transformQueue = transformData.BufferedTransform[i];
-                if (transformQueue.Length == 0)
-                {
-                    continue;
-                }
+                var transformBuffer = data.TransformBuffer[i];
+                var lastTransformApplied = data.LastTransformValue[i].Transform;
 
-                var nextTransform = transformQueue[0].transformUpdate;
-                
-                if (!tickOffsetSet)
+                // Need to take smear into account here when it's turned on
+                if (transformBuffer.Length == TransformSynchronizationConfig.MaxLoadMatchedBufferSize)
```

Comment:
Actually yes. This is because this isn't how this system is going supposed to work, but I left it like this until we turn the smearing back on. (Also that condition is the rarely hit one which is why it hasn't appeared in testing.